BUG: Make bool(str_scalar) and str_scalar.astype(bool) consistent

eric-wieser opened the pull request:
Fixes #9847, by changing `str_scalar.astype(bool)` to no longer mean `bool(int(str_scalar))`

eric-wieser commented on doc/release/1.14.0-notes.rst:
```diff
@@ -136,6 +136,14 @@ the result is always a view on the original masked array.
+``string.astype(bool)`` now uses the normal python semantics for converting
+---------------------------------------------------------------------------
+Previously this was interpreted as ``string.astype(int).astype(bool)``, which
+interpeted ``'0'`` as ``False``.
```
Do we want to go through a `FutureWarning` here?

mhvk commented:
Failures seem real. Not sure I understand why `getitem` is no good, so am not much use as a reviewer...

eric-wieser commented:
I wish `runtests.py` would catch the warnings that cause travis to fail like that. Hopefully fixed now.

ahaldane commented:
Another way the old behavior was bad:  `np.array(['abc']).astype(bool)` fails with a nonsensical error `invalid literal for int() with base 10`.

ahaldane commented:
This seems like a fairly clear "good move", and code looks fine to me.

As to whether we want a round of deprecation with FutureWarning: I lean towards yes, since it's not a very critical bug, yet the change is quite user-visible and I can even imagine someone depending on it. So I imagine a FutureWarning for 1.14, then change the behavior in 1.15.

ahaldane commented:
One thing that is arguably worse this way is that this no longer works:
```python
>>> np.array([True, False], dtype=bool).astype(str).astype(bool)
```
but I'm not sure it should work. After all, `bool(str(bool(False)))` is `True` in Python.

eric-wieser commented:
> I lean towards yes, since it's not a very critical bug, yet the change is quite user-visible and I can even imagine someone depending on it.

A naive implementation of adding the following to `numpy\core\src\multiarray\arraytypes.c.src:1577` 
```
#if defined(IS_TO_BOOL)
            /* Numpy 1.14, 2017-10-11 */
            if (DEPRECATE_FUTUREWARNING(
                "In future, casting string-like types to bool will not cast "
                "to int first. To keep the current behaviour of having strings "
                "of '0's be considered false, use .astype(int).astype(bool)."
                ) < 0) {
                return;
            }
#endif
```

\leads to the warning being emitted once for every single element in the array!

```
>>> a = np.array(list('test'))
>>> a.astype(bool)
FutureWarning: In future, casting string-like types to bool will not cast to int first. To keep the current behaviour of having strings of '0's be considered false, use .astype(int).astype(bool).
FutureWarning: In future, casting string-like types to bool will not cast to int first. To keep the current behaviour of having strings of '0's be considered false, use .astype(int).astype(bool).
FutureWarning: In future, casting string-like types to bool will not cast to int first. To keep the current behaviour of having strings of '0's be considered false, use .astype(int).astype(bool).
FutureWarning: In future, casting string-like types to bool will not cast to int first. To keep the current behaviour of having strings of '0's be considered false, use .astype(int).astype(bool).
array([ True,  True,  True,  True], dtype=bool)
```

Should I only emit the warning if the behaviour will change? That's still pathological for `np.array(['0'] * 10000)`

eric-wieser committed:
BUG: Make bool(str_scalar) and str_scalar.astype(bool) consistent

Caused by string values being cast to an int first and then a bool, ie arr.astype(int).astype(bool).
This causes crashes on non-numeric strings, and only accepts strings of 0 as false.

Fixes #9847

charris commented:
Needs rebase, also to get the merged #9856.

eric-wieser commented:
Rebased, but not ready to merge without a decision about if and when to `FutureWarning`

eric-wieser commented:
Closing in favor of having discussion in #9875, since this is a wider issue than anticipated
